text/0001-poet2-consensus.md

```diff
@@ -0,0 +1,109 @@
+
+## Block Publishing
+With the ClaimBlock ready and the WaitCertificate created, the validator will determine the duration for which to set the timer. The duration of the wait is determined by the ‘Duration’ field in the WaitCertificate.
+Ideally, any selected algorithm should be capable of converting the duration to a wait time distributed uniformly between 0 and LocalMean seconds.
+
+## Block Verification
+Upon receiving a claim block and a WaitCertificate, the receiving validator does the following:
+### Sanity checks
+1.	Ensure the Block and WaitCertificate signatures are valid
+2.	The Block’s ID matches the WaitCert’s blockID
+3.	If the block extends the current chain head, proceed to the Block Validity check. If it is a ‘future’ block, cache it for processing later.
+4.	If the block claims a previous block as the parent, proceed to ‘Fork Resolution’
+### Validity check & block publishing
+5.	Compute the new CC. The block is valid if CC + E < WC. Here E is a network latency coefficient added to simulate the average network delay seen by the validator.
+6.	Send the block to the validator for processing the transactions
+7.	If valid, commit the block, store the blockID and WaitCert in SGX
+8.	Forward the block to the gossip network
+9.	Drop the existing block proposal and restart from the new chain head.
+### Early Arriving blocks
+10.	If the block is invalid i.e. it is an early-arriving block, compare the duration with the duration of the claim block. If the block duration is smaller, drop the claim block. Wait for the WC to catch up with the CC. If the claim block duration is less than the incoming block duration, discard the incoming block.
+11.	Proceed to step 6 once the block becomes valid.
+### Fork Resolution
+12.	If the new block claims an earlier block as a parent, check for any cached blocks claiming the new block as a parent, retrieve them. Compute the new CC and check if the chain is valid (CC + E < WC)
+13.	If valid, compare existing chain length with the new chain length. If existing chain length is more, discard new chain. If the new chain is longer, switch to the new chain (commit the new blocks)
+14.	If the two chains are equal, compare Chain Clocks. Select the chain with the smaller CC.
+15.	If the chain is switched, drop the block proposal & restart (Step 9).
+16.	Forward the new block over the gossip network.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+Poet 2.0 improves on PoET 1.0 adding support for PSE free devices. Current PoET 1.0 limits the availability of PoET consensus only to SGX PSE enabled devices.
```

Comment:
As with my first comment above, I suggest using the term "(SGX) devices with/without Platform Services" (or something similar), instead of "PSE-free" or "PSE-enabled".